internal/services/oidc.go

```diff
@@ -0,0 +1,206 @@
+	sessionStorage *storages.SessionStorage) *OIDCService {
+	return &OIDCService{
+		client:         client,
+		sessionStorage: sessionStorage,
+		logger:         logger,
+	}
+}
+
+// GetOIDCProviderURL gets OIDC provider URL from the OAuth2 configuration.
+func (osvc *OIDCService) GetOIDCProviderURL() (string, error) {
+	// Creates a temporary session with state and nonce to validate
+	// when there is callback from the OIDC Provider.
+	session, err := models.CreateSession()
+	if err != nil {
+		return "", err
+	}
+
+	osvc.sessionStorage.StoreSession(session)
+
+	return osvc.client.OAuth2Config.AuthCodeURL(
+		session.State,
+		oidc.Nonce(session.Nonce),
+	), nil
+}
+
+// ClaimsToken creates claims token.
+func (osvc *OIDCService) ClaimsToken(ctx context.Context, code string) (models.Tokens, error) {
+	if code == "" {
+		return models.Tokens{}, errors.New("unable to authenticate without code returned from OIDC Provider")
+	}
+
+	oauth2Token, err := osvc.client.OAuth2Config.Exchange(ctx, code)
+	if err != nil {
+		return models.Tokens{}, errors.New("failed to exchange token. Error: " + err.Error())
+	}
+	idToken, err := osvc.validateIDToken(ctx, oauth2Token)
+	if err != nil {
+		return models.Tokens{}, err
+	}
+
+	idTokenClaims := new(json.RawMessage)
+	if err := idToken.Claims(&idTokenClaims); err != nil {
+		return models.Tokens{}, errors.New("An error occurred while validating ID Token. Error: " + err.Error())
+	}
+
+	userInfo, err := osvc.getUserInfo(ctx, oauth2Token)
+	if err != nil {
+		return models.Tokens{}, errors.New("Failed to get UserInfo. Error: " + err.Error())
+	}
+
+	cc := models.CustomClaims{}
+	if err := idToken.Claims(&cc); err != nil {
+		return models.Tokens{}, errors.New("An unexpected error has occurred. Error: " + err.Error())
+	}
+
+	return models.Tokens{
+		OAuth2Token:   oauth2Token,
+		IDTokenClaims: idTokenClaims,
+		UserInfo:      userInfo,
+		CustomClaims:  cc,
+	}, nil
+}
+
+// IsFlowSecure ensures the flow is secure and then, returns a JWT token.
+func (osvc *OIDCService) IsFlowSecure(state string, token models.Tokens) (bool, error) {
+	if state == "" {
+		return false, errors.New("unable to follow a secure flow without the state returned from the OIDC Provider")
+	}
+
+	session, err := osvc.sessionStorage.GetSession(state)
+	if err != nil {
+		return false, err
+	}
+
+	nonceToken, err := osvc.getValueFromToken("nonce", token)
+	if err != nil {
+		return false, err
+	}
+	if !session.MatchNonce(nonceToken.(string)) {
+		return false, errors.New("nonce did not match")
+	}
+
+	return true, nil
+}
+
+// CreateJSONData creates json data to display as body content.
+func (osvc *OIDCService) CreateJSONData(token models.Tokens) ([]byte, error) {
+	jwt, err := osvc.createJWT(&token.CustomClaims)
+	if err != nil {
+		return nil, errors.New("an error occurred while creating a JWT. Error: " + err.Error())
+	}
+
+	// If all goes well, then create the JSON data to display as body content.
+	jsonBody, err := json.Marshal(jwt)
```

Comment:
we should return the jwt from here.
let's imagine you have more than one controller, one that is not expecting json but protobuf, for example, you will need to change this.
encoding and decoding is controller responsibility, it's the one who knows how to do it

Reply:
great! I'll change it.